core/frontend/src/Marker.ts

```diff
@@ -443,3 +457,106 @@ export abstract class MarkerSet<T extends Marker> {
     }
   }
 }
+
+/**
+ * A [[MarkerSet]] sub-class that uses a gready clustering algorithm. Sub-classes should use the [[getAverageLocation]] function
+ * to set the location of the marker they return in `getClusterMarker`.
+ * @public
+ */
+export abstract class GreedyClusteringMarkerSet<T extends Marker> extends MarkerSet<T> {
+  /// The radius (in pixels) for clustering markers, default 0 which implies calculating the radius based on the visible marker imageSize/size.
+  protected clusterRadius = 0;
+
+  /**
+   * Gets the average of the cluster markers worldLocation.
+   * @param cluster Cluster to get average location of.
+   * @returns The average of the cluster markers worldLocation.
+   */
+  protected getAverageLocation(cluster: Cluster<T>) {
+    const location = Point3d.createZero();
+    cluster.markers.forEach((marker) => location.addInPlace(marker.worldLocation));
+    location.scaleInPlace(1 / cluster.markers.length);
+    return location;
+  }
+
+  /**
+   * Sets the cluster's rect by averaging the rects of all the markers in the cluster.
+   * @param cluster Cluster to set rect of.
+   */
+  protected setClusterRectFromMarkers<T extends Marker>(cluster: Cluster<T>) {
+    const len = cluster.markers.length;
+    if (len > 1) {
+      const midPoint = Point3d.createZero();
+      const size = Point3d.createZero();
+      cluster.markers.forEach((marker) => {
+        const center = new Point2d(marker.rect.left + (marker.rect.width / 2), marker.rect.top + (marker.rect.height / 2));
```

Comment:
probably should make a temporary outside the loop so you don't create per-marker garbage.

Reply:
I removed the unnecessary new point and moved this whole function to the Cluster class.